File: Leetcode3546. Equal Sums grid Partition I/Solution.cpp
class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size(), n = grid[0].size();
        long long total = 0;
        
        for (auto &row : grid)
            for (int x : row)
                total += x;
        
        if (total % 2) return false;
        
        long long target = total / 2, sum = 0;
        
        for (int i = 0; i < m - 1; i++) {
            for (int j = 0; j < n; j++)
                sum += grid[i][j];
            if (sum == target) return true;
        }
        
        sum = 0;
        
        for (int j = 0; j < n - 1; j++) {
            for (int i = 0; i < m; i++)
                sum += grid[i][j];
            if (sum == target) return true;
        }
        
        return false;
    }
};


// runtime - 0ms
class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        int m = grid.size();
        int n = grid[0].size();

        long long total = 0;

        // total sum
        for (auto &row : grid) {
            for (auto &val : row) {
                total += val;
            }
        }

        // if odd → impossible
        if (total % 2 != 0) return false;

        long long target = total / 2;

        // 🔹 check horizontal cut
        long long curr = 0;
        for (int i = 0; i < m - 1; i++) {  // m-1 because last row not allowed
            for (int j = 0; j < n; j++) {
                curr += grid[i][j];
            }
            if (curr == target) return true;
        }

        // 🔹 check vertical cut
        curr = 0;
        for (int j = 0; j < n - 1; j++) {  // n-1 because last column not allowed
            for (int i = 0; i < m; i++) {
                curr += grid[i][j];
            }
            if (curr == target) return true;
        }

        return false;
    }
};



// runtime - 1ms
class Solution {
public:

    bool canPartitionGrid(vector<vector<int>>& grid) {
        int M = grid.size();
        int N = grid[0].size();
        long long totalSum = 0;
        vector<long long> rowSum(M, 0);
        vector<long long> colSum(N, 0);
        for(int x = 0; x < M; x++)
        {
            for(int y = 0; y < N; y++)
            {
                totalSum += grid[x][y];
            }
        }
            if(totalSum % 2 != 0)
        {
            return false;
        }
        long long halfSum = totalSum / 2;
        long long currentRow = 0;

        for(int x = 0; x < M; x++)
        {
            for(int y = 0; y < N; y++)
            {
                currentRow += grid[x][y];
                totalSum += grid[x][y];
            }
            if(currentRow == halfSum)
            {
                return true;
            }
            rowSum[x] = currentRow;
        }
                // We can't partition an odd total sum into two separate parts
        long long currentCol = 0;
        for(int y = 0; y < N; y++)
        {
            for(int x = 0; x < M; x++)
            {
                currentCol += grid[x][y];
                if(currentCol == halfSum)
                {
                    return true;
                }
            }

            colSum[y] = (currentCol);
        }


        for(const auto& sum : rowSum)
        {
            if (sum == halfSum)
            {
                return true;
            }
        }
        for(const auto& sum : colSum)
        {
            if (sum == halfSum)
            {
                return true;
            }
        }
        return false;

    }
};



// runtime - 2ms
class Solution {
public:
    bool canPartitionGrid(vector<vector<int>>& grid) {
        
        int n = grid.size();
        int m = grid[0].size();

        vector<long long> row(n,0), col(m,0);

        // compute row and column sums
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                row[i] += grid[i][j];
                col[j] += grid[i][j];
            }
        }

        // check horizontal cut
        long long total = 0;
        for(auto x : row) total += x;

        long long prefix = 0;

        for(int i=0;i<n-1;i++){
            prefix += row[i];
            if(prefix == total - prefix)
                return true;
        }

        // check vertical cut
        total = 0;
        for(auto x : col) total += x;

        prefix = 0;

        for(int i=0;i<m-1;i++){
            prefix += col[i];
            if(prefix == total - prefix)
                return true;
        }

        return false;
    }
};



// runtime - 3ms
class Solution {
public:

    bool canPartitionGrid(vector<vector<int>>& grid) {
        long long sum = 0;
        int n = grid.size();
        int m = grid[0].size();
        for(int i=0;i<n;i++)
            for(int j = 0;j<m;j++) sum += grid[i][j];
        
        if(sum %2) return false;
        long long curr = 0;
        for(int i=0;i<n;i++){
            for(int j=0;j<m;j++){
                curr += grid[i][j];
            }
            if(sum/2 == curr) return true;
        }
        //
        curr = 0;
        for(int j=0;j<m;j++){
            for(int i=0;i<n;i++){
                curr += grid[i][j];
            }
            if(sum/2 == curr) return true;
        }
        return false;
    }
};

File: Leetcode3546. Equal Sums grid Partition I/Solution.py
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        m, n = len(grid), len(grid[0])
        total = sum(sum(row) for row in grid)
        
        if total % 2:
            return False
        
        target = total // 2
        s = 0
        
        for i in range(m - 1):
            s += sum(grid[i])
            if s == target:
                return True
        
        s = 0
        
        for j in range(n - 1):
            for i in range(m):
                s += grid[i][j]
            if s == target:
                return True
        
        return False



# runtime - 44ms
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:

                # Sum rows of the grid or its transpose and 
                # determine whether the equal partition exists
        def partitionExists(arr: list, sm = 0)-> bool:

            for row in arr:
                sm+= sum(row)
                if sm == halfSum: return True
                if sm >  halfSum: return False


                # Determine the target sum for each partition
                # and whether sum(grid) is odd
        halfSum, mod_ = divmod(sum(chain(*grid)), 2)
        if mod_ : return False

                # Check horizontal and vertical partitions
        return (partitionExists(grid) or 
                partitionExists(zip(*grid)))   



# runtime - 48ms
from typing import List

class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        def is_valid(sums: List[int]) -> bool:
            total = sum(sums)
            if total % 2 != 0:
                return False
            
            target = total//2
            running_total = 0
            for i in sums:
                running_total += i
                if running_total == target:
                    return True
            return False
            

        # Use lists to preserve duplicate sum values
        row_sums = [sum(row) for row in grid]
        col_sums = [sum(col) for col in zip(*grid)]
        
        return is_valid(row_sums) or is_valid(col_sums)


# runtime - 52ms
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        totalSum = sum(map(sum, grid))
        def canPartition(grid:list[list[int]]) -> bool:
            runningSum = 0
            for row in grid:
                runningSum += sum(row)
                if runningSum * 2 == totalSum:
                    return True
            return False
        return canPartition(grid) or canPartition(zip(*grid))



# runtime - 56ms
class Solution:
    def canPartitionGrid(self, G: List[List[int]]) -> bool:
        def checkRows(G):
            m, n = len(G), len(G[0])
            rows = [sum(G[i]) for i in range(m)]
            total = sum(rows)
            prefixSum = 0
            for row in rows:
                prefixSum += row
                if prefixSum == total - prefixSum:
                    return True
            return False
        
        if checkRows(G):
            return True
        GT = list(zip(*G))
        return checkRows(GT)


# runtime - 60ms
class Solution:
    def canPartitionGrid(self, grid: List[List[int]]) -> bool:
        s = sum(sum(row) for row in grid)
        if s % 2:
            return False
        pre = 0
        for i, row in enumerate(grid):
            pre += sum(row)
            if pre * 2 == s and i != len(grid) - 1:
                return True
        pre = 0
        for j, col in enumerate(zip(*grid)):
            pre += sum(col)
            if pre * 2 == s and j != len(grid[0]) - 1:
                return True
        return False
